### PR TITLE
add Memory Manager configuration flag

### PR DESCRIPTION
**Pull request by bg-chun**
ToDoList:
- add unittest for validation logic on server.go
  => Done
- add validation logic on memorymanager.go
    (It seems #15 is opened for this)

### CODE REVIEW

File: cmd/kubelet/app/server.go [hunk 1]
```diff
@@ -1243,6 +1251,60 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
...
+			}
+		}
+
+		limitInString, _ := m[limitKey]
+		limit, err := resource.ParseQuantity(limitInString)
+		if err != nil || limit.Sign() != -1 {
+			return nil, fmt.Errorf("memory limit conversion error for value \"%s\"", limitInString)
+		}
+
+		parsed[idx][v1Type] = limit
```
Comment (k-wiatrzyk) [inline cmd/kubelet/app/server.go:1301]: A nested map is not initialized: `parsed[idx] = make(map[v1.ResourceName]resource.Quantity)`

File: cmd/kubelet/app/server.go [hunk 2]
```diff
@@ -1243,6 +1251,60 @@ func parseResourceList(m map[string]string) (v1.ResourceList, error) {
...
+		if corev1helper.IsHugePageResourceName(v1Type) {
+			if _, err := corev1helper.HugePageSizeFromResourceName(v1Type); err != nil {
+				return nil, fmt.Errorf("memory type conversion error, unknown type: \"%s\"", typeInString)
+			}
+		}
+
+		limitInString, _ := m[limitKey]
+		limit, err := resource.ParseQuantity(limitInString)
+		if err != nil || limit.Sign() != -1 {
+			return nil, fmt.Errorf("memory limit conversion error for value \"%s\"", limitInString)
```
Comment (k-wiatrzyk) [inline cmd/kubelet/app/server.go:1298]: Without minus? `limit.Sign() != 1`

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (bg-chun) [reply cmd/kubelet/app/server.go:1301]: yeah I just got an error on unittest due to this.
```
=== RUN   TestValueOfPreReservedMemoryZone
--- FAIL: TestValueOfPreReservedMemoryZone (0.00s)
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map
```

File: cmd/kubelet/app/server.go (hunk 2 above)
Comment (bg-chun) [reply cmd/kubelet/app/server.go:1298]: you are right, I think I got destracted by something.
```
// Sign returns 0 if the quantity is zero, -1 if the quantity is less than zero, or 1 if the
// quantity is greater than zero.
func (q *Quantity) Sign() int {
```

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (k-wiatrzyk) [reply cmd/kubelet/app/server.go:1301]: Oh, I also was checking it with unit tests. The test function is in PR #17 , if you did it too then I will close my PR.

File: cmd/kubelet/app/server.go (hunk 2 above)
Comment (bg-chun) [reply cmd/kubelet/app/server.go]: Done

File: cmd/kubelet/app/server.go (hunk 1 above)
Comment (bg-chun) [reply cmd/kubelet/app/server.go]: Oh I didnt noticed that...
I uploaded commit for unittest.
https://github.com/mgjeong/kubernetes/pull/16/commits/6272a3c6f083b79438b71c0f581a2c509f77d788

Comment (bg-chun) [reply cmd/kubelet/app/server.go]: I think cherry-pick your commit here is better let me try

Comment (bg-chun) [reply cmd/kubelet/app/server.go]: Done

File: cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -688,6 +689,12 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, featureGate f
 			s.SystemReserved["cpu"] = strconv.Itoa(reservedSystemCPUs.Size())
 			klog.Infof("After cpu setting is overwritten, KubeReserved=\"%v\", SystemReserved=\"%v\"", s.KubeReserved, s.SystemReserved)
 		}
+
+		preReservedMemoryZone, err:= parsePreReservedMemoryConfig(s.PreReservedMemoryZone)
```
Comment (cezaryzukowski) [inline cmd/kubelet/app/server.go:693]: space after `err`?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+package flag
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+)
+
+// BracketSeparatedSliceMapStringString can be set from the command line with the format `--flag {key=value, ...}, {...}`.
+// Multiple comma-separated key-value pairs in a braket(`{}`) in a single invocation are supported. For example: `--flag {key=value, key=value, ...}`.
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go:26]: `bracket`?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+// Multiple braket-separated list of key-value pairs in a single invocation are supported. For example: `--flag {key=value, key=value}, {key=value, key=value}`.
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go:27]: `bracket`?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+		m.initialized = true
+	}
+
+	value = strings.TrimSpace(value)
+
+	// split here
+	//{numa-node=0,memory-type=memory,limit=1Gi},{numa-node=1,memory-type=memory,limit=1Gi},{numa-node=1,memory-type=memory,limit=1Gi}
+//	for _, split := range strings.Split(value, "{") {
+//		split = strings.TrimRight(split, ",")
+//		split = strings.TrimRight(split, "}")
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go:56]: Should we remove those comments?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	for _, split := range strings.Split(value, ",{") {
+		//split = strings.TrimRight(split, ",")
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string.go:58]: Should we remove?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+)
+
+func TestStringBracketSeparatedSliceMapStringString(t *testing.T) {
+	var nilSliceMap []map[string]string
+	testCases := []struct {
+		desc   string
+		m      *BracketSeparatedSliceMapStringString
+		expect string
+	}{
+		{"nill", NewBracketSeparatedSliceMapStringString(&nilSliceMap), ""},
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go:31]: `nil`?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+		err    string
+	}{
+		// we initialize the map with a default key that should be cleared by Set
+		{"clears defaults", []string{""},
+			NewBracketSeparatedSliceMapStringString(&[]map[string]string{{"default": ""}}),
+			&BracketSeparatedSliceMapStringString{
+				initialized: true,
+				Value:       &[]map[string]string{},
+			}, ""},
+		// make sure we still allocate for "initialized" multimaps where Multimap was initially set to a nil map
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go:62]: 👍

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+				Value:       &[]map[string]string{},
+			}, ""},
+		// for most cases, we just reuse nilMap, which should be allocated by Set, and is reset before each test case
+		{"empty", []string{""},
+			NewBracketSeparatedSliceMapStringString(&nilMap),
+			&BracketSeparatedSliceMapStringString{
+				initialized: true,
+				Value:       &[]map[string]string{},
+			}, ""},
+		{"empty braket", []string{"{}"},
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go:76]: `bracket`?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+			NewBracketSeparatedSliceMapStringString(&nilMap),
+			&BracketSeparatedSliceMapStringString{
+				initialized: true,
+				Value:       &[]map[string]string{},
+			}, ""},
+		{"missing braket", []string{"a=string, b=string"},
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go:82]: `bracket`?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+				initialized: true,
+				Value:       &[]map[string]string{{"a" : "string"}},
+			}, ""},
+		{"two keys", []string{"{a=string,b=string}"},
+			NewBracketSeparatedSliceMapStringString(&nilMap),
+			&BracketSeparatedSliceMapStringString{
+				initialized: true,
+				Value:       &[]map[string]string{{"a" : "string", "b" : "string"}},
+			}, ""},
+		{"two duplecated keys", []string{"{a=string,a=string}"},
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go:106]: `duplicated`?

File: staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+			NewBracketSeparatedSliceMapStringString(&nilMap),
+			&BracketSeparatedSliceMapStringString{
+				initialized: true,
+				Value:       &[]map[string]string{{"a" : "string"}},
+			}, ""},
+		{"two keys with space", []string{"{a = string, b = string}"},
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/component-base/cli/flag/bracket_separated_slice_map_string_string_test.go:112]: `spaces`?

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go [hunk 14]
```diff
@@ -765,6 +765,14 @@ type KubeletConfiguration struct {
 	// Default: []
 	// +optional
 	AllowedUnsafeSysctls []string `json:"allowedUnsafeSysctls,omitempty"`
+	// A braket separated list of configuration for mamory manager.
```
Comment (cezaryzukowski) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:768]: `A comma separated list of bracket-enclosed configuration for memory manager`?

**Pull request merged**